gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -296,9 +300,68 @@ def __init__(self,
               self._process_header_lines),
           **kwargs)
 
+    if isinstance(self, PySamParser):
+      # In an absence of "#CHROM POS..." header line, PySam process would lock.
+      # However, we need to support cases when no records in files were provided
+      # even if the above header line is missing. To circumvent the error, read
+      # the file until the first line is found, and verify validity.
+      self._header_lines = None
+      if isinstance(range_tracker, range_trackers.UnsplittableRangeTracker):
+        secondary_tracker = range_trackers.UnsplittableRangeTracker(
+            copy(range_tracker._range_tracker))
+      else:
+        secondary_tracker = copy(range_tracker)
+      self._verify_header(copy(text_source), secondary_tracker)
+
     self._text_lines = text_source.read_records(self._file_name,
                                                 range_tracker)
 
+  def after_read(self):
+    return
+
+  def _next_non_empty_line(self, iterator):
+    # Get next non-empty stripped record from iterator.
+    # type: (typing.Iterable[str]) -> str
+    text_line = next(iterator).strip()
+    while not text_line:  # skip empty lines.
+      text_line = next(iterator).strip()
+    return text_line
+
+  def _verify_header(self, text_source, range_tracker):
+    try:
+      text_lines = text_source.read_records(self._file_name,
+                                            range_tracker)
+      # Read until first non empty record
+      self._next_non_empty_line(text_lines)
+      if not self._header_lines[-1].startswith(LAST_HEADER_LINE_PREFIX):
+        raise ValueError('Header missing CHROM line.')
+      # Construct VariantHeader class from the metadata lines, to verify the
+      # validity of the file. PySam would not raise an error, will omit the
+      # metadata and hide the issue, if the header is incorrect.
+      header = libcbcf.VariantHeader()
+      for header_line in self._header_lines[:-1]:
+        header.add_line(header_line)
+
+      for k, v in header.info.iteritems():
```

Comment:
Does PyVCF raise errors for these cases? I assume PySam does not so we have to check it by ourselves?

Reply:
Yeah, PyVCF does, there are tests for that.

PySam is too smart for its own good, and tries to swallow this issues to allow clients to go through even with incorrect headers. However, for our case, we would want to report header issues upstream to the caller.